src/descriptor.rs

```diff
@@ -0,0 +1,567 @@
+use std::{
+    collections::HashMap,
+    mem::{size_of, transmute},
+};
+
+#[repr(u8)]
+#[derive(num_enum::FromPrimitive, num_enum::IntoPrimitive, Copy, Clone, PartialEq, Eq, Debug)]
+pub(crate) enum DescriptorType {
+    #[num_enum(default)]
+    Other = 0x00,
+    Device = 0x01,
+    Configuration = 0x02,
+    Interface = 0x04,
+    Endpoint = 0x05,
+}
+
+/// All standard descriptors have these 2 fields in common
+#[derive(Copy, Clone, Debug)]
+#[repr(C, packed)]
+struct DescriptorHeader {
+    bLength: u8,
+    bDescriptorType: u8,
+}
+
+impl DescriptorHeader {
+    fn from_slice(slice: &[u8]) -> Result<Self, ParseError> {
+        if slice.len() < size_of::<Self>() {
+            return Err(ParseError);
+        }
+
+        let bytes: [u8; size_of::<Self>()] = slice[..size_of::<Self>()].try_into().unwrap();
+        // safety: self is valid for all bit patterns.
+        let res: Self = unsafe { transmute(bytes) };
+
+        // avoid infinite loop when bLength = 0
+        if (res.bLength as usize) < size_of::<Self>() {
+            return Err(ParseError);
+        }
+
+        Ok(res)
+    }
+}
+
+#[derive(Copy, Clone, Debug)]
+#[repr(C, packed)]
+pub(crate) struct DeviceDescriptor {
+    pub(crate) bLength: u8,
+    pub(crate) bDescriptorType: u8,
+    pub(crate) bcdUSB: u16,
+    pub(crate) bDeviceClass: u8,
+    pub(crate) bDeviceSubClass: u8,
+    pub(crate) bDeviceProtocol: u8,
+    pub(crate) bMaxPacketSize0: u8,
+    pub(crate) idVendor: u16,
+    pub(crate) idProduct: u16,
+    pub(crate) bcdDevice: u16,
+    pub(crate) iManufacturer: u8,
+    pub(crate) iProduct: u8,
+    pub(crate) iSerialNumber: u8,
+    pub(crate) bNumConfigurations: u8,
+}
+
+#[derive(Copy, Clone, Debug)]
+#[repr(C, packed)]
+struct ConfigurationDescriptor {
+    bLength: u8,
+    bDescriptorType: u8,
+    wTotalLength: u16,
+    bNumInterfaces: u8,
+    bConfigurationValue: u8,
+    iConfiguration: u8,
+    bmAttributes: u8,
+    bMaxPower: u8,
+}
+
+#[derive(Copy, Clone, Debug)]
+#[repr(C, packed)]
+struct InterfaceDescriptor {
+    bLength: u8,
+    bDescriptorType: u8,
+    bInterfaceNumber: u8,
+    bAlternateSetting: u8,
+    bNumEndpoints: u8,
+    bInterfaceClass: u8,
+    bInterfaceSubClass: u8,
+    bInterfaceProtocol: u8,
+    iInterface: u8,
+}
+
+#[derive(Copy, Clone, Debug)]
+#[repr(C, packed)]
+struct EndpointDescriptor {
+    bLength: u8,
+    bDescriptorType: u8,
+    bEndpointAddress: u8,
+    bmAttributes: u8,
+    wMaxPacketSize: u16,
+    bInterval: u8,
+    bRefresh: u8,
+    bSynchAddress: u8,
+}
+
+#[derive(Copy, Clone, Debug)]
+#[repr(C, packed)]
+struct BosDescriptor {
+    bLength: u8,
+    bDescriptorType: u8,
+    wTotalLength: u16,
+    bNumDeviceCaps: u8,
+}
+
+/// Descriptors provided by the USB device are malformed.
+#[derive(Copy, Clone, PartialEq, Eq, Debug)]
+pub(crate) struct ParseError;
+
+/// safety: self must be valid for all bit patterns, must not have padding holes.
+pub(crate) unsafe trait Descriptor: Sized {
```

Comment:
Or `zerocopy`, because it has better support for dealing with endianness.

Reply:
i've added `.to_native_endian()` to handle endianness.

I didn't use bytemuck/zerocopy to avoid adding more dependencies. I find them somewhat heavy because they bring in all the proc macro stuff, and maybe too complex (they do a lot more stuff than we need). No strong opinion though, if you do prefer to use them let me know and I'll change it.

Reply:
Seems ok. This library is definitely not `#[forbid(unsafe_code)]`!